src/test/java/com/waffle/dredes/level/LevelGen.java

```diff
@@ -82,151 +113,199 @@ public Tile[][] generate(Biome biome, int x, int y, boolean river)
     }
 
     public void loadSand(RoomLoader roomLoader)
     {
         roomLoader.addDirectory("DreDes/Rooms/Hills");
-        roomLoader.addDirectory("DreDes/Rooms/Caves");
+        roomLoader.addDirectory("DreDes/Rooms/Cave");
         roomLoader.addDirectory("DreDes/Rooms/Pit");
         roomLoader.addDirectory("DreDes/Rooms/Temple");
     }
 
     public void loadSalt(RoomLoader roomLoader)
     {
         roomLoader.addDirectory("DreDes/Rooms/Cliff");
-        roomLoader.addDirectory("DreDes/Rooms/Caves");
+        roomLoader.addDirectory("DreDes/Rooms/Cave");
         roomLoader.addDirectory("DreDes/Rooms/Pit");
         roomLoader.addDirectory("DreDes/Rooms/SaltFlats");
         roomLoader.addDirectory("DreDes/Rooms/Temple");
     }
 
     public void loadStone(RoomLoader roomLoader)
     {
-        roomLoader.addDirectory("DreDes/Rooms/Caves");
+        roomLoader.addDirectory("DreDes/Rooms/Cave");
         roomLoader.addDirectory("DreDes/Rooms/Mountain");
         roomLoader.addDirectory("DreDes/Rooms/FrozenPond");
         roomLoader.addDirectory("DreDes/Rooms/Temple");
     }
 
     public Room pickRoom(Room[][] rooms, int row, int col)
     {
-        HashSet<Room> pool = new HashSet<Room>();
-        try {
-            pool.addAll(Arrays.asList(rooms[row - 1][col - 1].getRoomPool(DOWN_RIGHT, true)));
-        }catch (Exception e){};
-        try {
-            pool.addAll(Arrays.asList(rooms[row][col - 1].getRoomPool(RIGHT, true)));
-        }catch (Exception e){};
-        try {
-            pool.addAll(Arrays.asList(rooms[row + 1][col - 1].getRoomPool(UP_RIGHT, true)));
-        }catch (Exception e){};
-        try {
-            pool.addAll(Arrays.asList(rooms[row - 1][col].getRoomPool(UP, true)));
-        }catch (Exception e){};
-        try {
-            pool.addAll(Arrays.asList(rooms[row + 1][col].getRoomPool(DOWN, true)));
-        }catch (Exception e){};
-        try {
-            pool.addAll(Arrays.asList(rooms[row - 1][col + 1].getRoomPool(DOWN_LEFT, true)));
-        }catch (Exception e){};
-        try {
-            pool.addAll(Arrays.asList(rooms[row][col + 1].getRoomPool(LEFT, true)));
-        }catch (Exception e){};
-        try {
-            pool.addAll(Arrays.asList(rooms[row][col + 1].getRoomPool(UP_LEFT, true)));
-        }catch (Exception e){};
-        if (!pool.isEmpty())
+        Collection<Room> pool = roomLoader.getRooms().values();
+        pool.remove(roomLoader.getRoom("Source"));
+        pool.remove(roomLoader.getRoom("Camp"));
+        for(int i = -1; i < 2; i++)
+        {
+            for(int j = -1; j < 2; j++)
+            {
+                if(row + i > 0 && row + 1 < rooms.length && col + j > 0 && col + j < rooms[0].length && i != 0 && j != 0)
+                {
+                    if(rooms[row + i][col + j] != null)
+                    {
+                        if(!Arrays.asList(rooms[row + i][col + j].getRoomPool(Room.getDirection(new Vec2f(-j, -i)), true)).isEmpty())
```

Comment:
For longer if statement checks like this, you may want to define a function for _what_ the if statement is trying to do. Also, consider [inverting your ifs](https://www.jetbrains.com/help/resharper/InvertIf.html)